### PR TITLE
native-client-software-grant - ClientMetadata::getServerLocation is r…

### PR DESCRIPTION
**Pull request by fdaniel7**
…eturning a primary BucketServerLocationPtr instead of a secondary

A pointer to a primary BucketServerLocationPtr can be returned when
asking for a secondary. The returned BucketServerLocationPtr is a
randomization between [0,
m_bucketServerLocationsList[bucketId].size()-1] where the result can be
a primary copy.

### CODE REVIEW

**Comment by PivotalSarge** [comment]
From inspection of the places where ClientMetadata::getServerLocation() gets called, I do not believe that it ought to return only secondaries when tryPrimary == false. Additionally, setting serverLocation to NULL when the list of locations is empty may break assumptions by the caller. I think there should be more information about why it is needed since ClientMetadata.hpp is not a public API and more analysis into how this change it would not break the existing calls in ClientMetaService.

**Comment by fdaniel7** [comment]
It looks like when tryPrimary == false, the intention was that ClientMetadata::getServerLocation() will return a secondary bucket from m_bucketServerLocationsList[bucketId], considering that there's only one primary location and 0 or more secondaries. the range of the randomization is definitely wrong in m_bucketServerLocationsList[bucketId].at(randgen((int)m_bucketServerLocationsList[bucketId].size())). Since we don't want to assume that the primary is always located in m_bucketServerLocationsList[bucketId].at(0), a while loop is added.
I don't see a point splitting this function to sections unless  we want specifically primary bucket or secondary.

The reason why I wanted to change this code is because  ClientMetadataService::markPrimaryBucketForTimeoutButLookSecondaryBucket() doesn't work properly since prBuckets->setBucketTimeout(i) will be called only if getServerLocation randomized the primary bucket, which makes no sense.
I'm going to add new pull request with a new function for marking buckets as blacklisted.

File: geode-client-native/src/cppcache/impl/ClientMetadataService.cpp [hunk 1]
```diff
@@ -1,925 +1 @@
...

  int32_t bucketSetSetSize = bucketSet->size();



  LOGDEBUG("ClientMetadataService::AreBucketSetsEqual currentBucketSetSize = %d bucketSetSetSize = %d ", currentBucketSetSize, bucketSetSetSize);



  if (currentBucketSetSize != bucketSetS
```
Comment (pivotal-jbarrett) [inline geode-client-native/src/cppcache/impl/ClientMetadataService.cpp]: It appears you have added new whitespace to every line in this file. Please correct.

Comment (fdaniel7) [inline geode-client-native/src/cppcache/impl/ClientMetadataService.cpp]: Hi, whitespaces removed

**Comment by doribd** [comment]
When you 'pivotal-jbarrett' said that fdaniel should close and switch the poll to the native.
You mean open it there and close this here right?

**Comment by echobravopapa** [comment]
@doribd yes, @fdaniel7 should close the PR on apache/geode/native-client-software-grant
and then switch to https://github.com/apache/geode-native and re-open the PR there as appropriate.

**Comment by doribd** [comment]
ok thanks, understood now.

**Comment by kohlmu-pivotal** [comment]
@doribd @fdaniel7 could you please close this PR if it is not required anymore.

**Comment by pivotal-jbarrett** [comment]
Closing for no activity.

**Pull request closed without merging**